### PR TITLE
Make test_extract take `&Path`s rather than `&str`s.

### PR DESCRIPTION
**Pull request by ltratt**
lang_tester tries to make it as easy as possible to set-up, so it has a fairly simple API. One way I did that was by passing the contents of a file to a `test_extract` function. However, that is fairly inflexible, and also it turns out not very much simpler (it doesn't help that `test_extract` returns `Option<String>`: frankly, it's easier for tests to panic in such situations).

This commit therefore changes the `test_extract` function signature from:

  ```
  Fn(&str) -> Option<String> + Send + Sync,
  ```

  to:

  ```
  Fn(&Path) -> String + Send + Sync,
  ```

In other words, users now have to a) read the contents of a path themselves (but
it doesn't necessarily have to be the path passed to the function!) b) return a
`String` rather than an `Option<String>`.

In practise, most `test_extract` functions can be changed from (roughly):

  ```
  test_extract(|s| { s.lines() ... })
  ```

to:

  ```
  test_extract(|p| { read_to_string(p).lines() })
  ```

Fixes https://github.com/softdevteam/lang_tester/issues/76 though at the expense of a relatively painful API bump. I think it's worth it though.

### CODE REVIEW

File: CHANGES.md [hunk 1]
```diff
@@ -1,3 +1,29 @@
...
+  Fn(&str) -> Option<String> + Send + Sync,
+  ```
+  to:
+  ```
+  Fn(&Path) -> String + Send + Sync,
+  ```
+
+  In other words, users now have to a) read the contents of a path themselves
+  (but it doesn't necessarily have to be the path passed to the function!) b)
+  return a `String` rather than an `Option<String>`.
```
Comment (jacob-hughes) [inline CHANGES.md:14]: Can we try rephrasing this? It took me a couple of re-reads to parse correctly as I read the "a)" and "b)" as mutually exclusive options. Even just switching to _"users now have to do two things: ...."_ would help.

**Comment by jacob-hughes** [comment]
Minor comment about phrasing in the README, otherwise LGTM. I think I prefer this API.

File: CHANGES.md (hunk 1 above)
Comment (ltratt) [reply CHANGES.md]: Is https://github.com/softdevteam/lang_tester/pull/77/commits/937290345004d94d4b080aa0e6f66002b397ac31 enough or is more needed?

Comment (jacob-hughes) [reply CHANGES.md]: Yes that's much better. Can you change it in the commit message too when you squash.

**Comment by jacob-hughes** [comment]
Pleas squash

File: CHANGES.md (hunk 1 above)
Comment (ltratt) [reply CHANGES.md]: Will do!

**Comment by ltratt** [comment]
I'd like to wait a day or so to see if @xldenis feels this solves the issue he raised. If it does, I'll squash (and fix the commit message at the same time).

**Comment by jacob-hughes** [comment]
Sure

**Comment by xldenis** [comment]
This doesn't solve the underlying issue of #76 in my opinion as it still has a `String` output type. 

The use case I'm envisioning is to have `.stdout` and `.stderr` alongside my tests and use those as the expected output. 

It seems like if I want to use this in the current api I need to construct a string that corresponds to the test DSL header out of those files? 

I see two real solutions: structure the output type, or expose a lower level primitive which asks users to directly construct a `Tests` struct.

**Comment by ltratt** [comment]
I think there's a fairly simple work around. Your `test_extract` function can read `x.stderr` and `x.stdout` and combine them into a `String` which looks as follows:

```
stdout:
  ...
stderr:
  ...
```

Does that sound doable?

**Comment by xldenis** [comment]
@ltratt it is doable but it feels like a less satisfactory api if I'm honest, why pass around unstructured types when we can have structure?

**Comment by xldenis** [comment]
What is the problem with using something like:

```
pub fn test_extract_prim<F>(&mut self, test_extract: F) -> &mut Self
    where
        F: 'static + Fn(&Path) -> Tests + Send + Sync,
```
?

The existing `test_extract` would just take a user closure returning a `String` wrap it, calling `parse_tests` and pass it along.

**Comment by ltratt** [comment]
@xldenis Structure comes at an ease-of-use cost: your API suggestion is nice in one way but now I need to expose `Tests` to the user and they have to understand it, and it increases the complexity. It's always a trade-off and lang_tester has been designed from the start to slightly favour ease-of-use over flexibility. However, I think your use case is something that lang_tester should be able support, hence this PR and the API change it necessitates. Your use case might not be supported in quite the way that you want, but at least we can offer you *something* now, which I hope is better than nothing :)

@jacob-hughes Squashed.

**Comment by xldenis** [comment]
Of course, I'll take this improvement! 

 I'm not convinced that an adhoc DSL is simpler than a structured type though, users still need to understand the structure and limitations of your parser.

**Comment by ltratt** [comment]
@xldenis There are definitely situations where lang_tester isn't sufficient: from the very beginning, I wanted to get 80% of the power of https://crates.io/crates/compiletest_rs with 20% of the complexity, and I think we're in that sort of range. There's probably a place for a much more flexible, but complex, language neutral equivalent of lang_tester / compiletest_rs, but we've been pleasantly surprised at how many projects we've been able to slot lang_tester into without issue.

**Comment by xldenis** [comment]
Fair enough! One final option before I let you go! Define a few `From` instances for `Tests` and use those as the output type.
You could have a `From<String>` which uses `parse_tests` and a `From<HashMap<String, (String, String)>` or whatever.

**Comment by ltratt** [comment]
Interesting idea. Let me chew on it!

**Comment by jacob-hughes** [comment]
@ltratt I don't think the commit message was updated?

**Comment by ltratt** [comment]
Mea culpa! Force pushed an update.
